Sources/WireGuardKitGo/in-tunnel-tcp.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import "C"
+
+import (
+	"context"
+	"net"
+	"net/netip"
+	"time"
+	"unsafe"
+
+	"golang.zx2c4.com/wireguard/tun/netstack"
+)
+
+// Opens a TCP connection to the specified address as though it was bound to the tunnel.
+// This function returns a socket handle immediately, and it can be used immediately after,
+// but the socket may not be connected immediately. When writing or reading from the socket,
+// the calls will wait until the socket connection is established or it times out.
+//
+//export wgOpenInTunnelTCP
+func wgOpenInTunnelTCP(tunnelHandle int32, address *C.char, timeout uint64) int32 {
+	tun := tunnels.Get(tunnelHandle)
+	if tun == nil {
+		return errNoSuchTunnel
+	}
+	if tun.VirtualNet == nil {
+		return errNoTunnelVirtualInterface
+	}
+
+	netAddr, err := netip.ParseAddrPort(C.GoString(address))
+	if err != nil {
+		return errBadIPString
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
+
+	connectTcpSocket := func(ctx context.Context, vnet *netstack.Net) (net.Conn, error) {
+		connection, err := vnet.DialContextTCPAddrPort(ctx, netAddr)
+		cancel()
+		return connection, err
+	}
+
+	return tun.AddSocket(ctx, connectTcpSocket)
+}
+
+//export wgCloseInTunnelTCP
+func wgCloseInTunnelTCP(tunnelHandle int32, socketHandle int32) bool {
+	tun := tunnels.Get(tunnelHandle)
+	if tun == nil {
+		return false
+	}
+
+	return tun.RemoveAndCloseSocket(socketHandle)
+}
+
+// Sends the data array into the TCP socket in a blocking fashion. The data
+// pointer should point to at least `dataLen` bytes for the entirety of this
+// call. This function is technically threadsafe, but multiple calls will not
+// have a defined order, which can lead to unordered writes.
+//
+//export wgSendInTunnelTCP
+func wgSendInTunnelTCP(tunnelHandle int32, socketHandle int32, data *byte, dataLen int32) int32 {
+	tun := tunnels.Get(tunnelHandle)
+	if tun == nil {
+		return errNoSuchTunnel
+	}
+
+	socket, err, ok := tun.GetSocket(socketHandle)
+	if !ok {
+		return errTCPNoSocket
+	}
+	if err != nil {
+		tun.logger.Errorf("Failed to open TCP socket: %s", err)
+		tun.RemoveAndCloseSocket(socketHandle)
+		return errTCPNoSocket
+	}
+
+	byteBuffer := C.GoBytes(unsafe.Pointer(data), C.int(dataLen))
```

Comment:
Why not use `unsafe.Slice(data, dataLen)` here?

Reply:
Done.
<!-- Reviewable comment -ODh6lLSBvblfEyZuOgR:b-896fix -->
<!-- Sent from Reviewable.io -->